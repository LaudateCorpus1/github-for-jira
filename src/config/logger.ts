import Logger, { levelFromName } from "bunyan";
import bformat from "bunyan-format";
import filteringStream from "../common/filteringStream";
import { LoggerWithTarget, wrapLogger } from "probot/lib/wrap-logger";
import { Request } from "express";


// For any Micros env we want the logs to be in JSON format.
// Otherwise, if local development, we want human readable logs.
const outputMode = process.env.MICROS_ENV ? "json" : "short";

// add levelInString to include DEBUG | ERROR | INFO | WARN
const formatOut = filteringStream(bformat({ outputMode, levelInString: true }));

const requestSerializer = (req: Request) => (!req || !req.connection) ? req : {
	method: req.method,
	url: req.originalUrl || req.url,
	path: req.path,
	headers: req.headers,
	remoteAddress: req.connection.remoteAddress,
	remotePort: req.connection.remotePort,
	body: req.body
};

const errorSerializer = (err) => (!err || !err.stack) ? err : {
	...err,
	response: Logger.stdSerializers.res(err.response),
	request: requestSerializer(err.request),
	stack: getFullErrorStack(err)
};

const getFullErrorStack = (ex) => {
	let ret = ex.stack || ex.toString();
	if (ex.cause && typeof (ex.cause) === "function") {
		const cex = ex.cause();
		if (cex) {
			ret += "\nCaused by: " + getFullErrorStack(cex);
		}
	}
	return ret;
};

const logger = Logger.createLogger(
	{
		name: "github-for-jira",
		logger: "root-logger",
		stream: formatOut,
		serializers: {
			err: errorSerializer,
			res: Logger.stdSerializers.res,
			req: requestSerializer
		}
	}
);

const logLevel = process.env.LOG_LEVEL || "info";
const globalLoggingLevel = levelFromName[logLevel] || Logger.INFO;
logger.level(globalLoggingLevel);

// TODO Remove after upgrading Probot to the latest version (override logger via constructor instead)
export const overrideProbotLoggingMethods = (probotLogger: Logger) => {
	// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// @ts-ignore

	// Remove  Default Probot Logging Stream
	probotLogger.streams.pop();

	// Replace with formatOut stream
	probotLogger.addStream({
		type: "stream",
		stream: formatOut,
		closeOnExit: false,
		level: globalLoggingLevel
	});
};

export const getLogger = (name: string): LoggerWithTarget => {
	return wrapLogger(logger.child({ logger: name }));
};

//Override console.log with bunyan logger.
//we shouldn't use console.log in our code, but it is done to catch
//possible logs from third party libraries
const consoleLogger = getLogger("console");
// eslint-disable-next-line no-console
console.debug = consoleLogger.debug.bind(consoleLogger);
// eslint-disable-next-line no-console
console.error = consoleLogger.error.bind(consoleLogger);
// eslint-disable-next-line no-console
console.log = consoleLogger.info.bind(consoleLogger);
// eslint-disable-next-line no-console
console.warn = consoleLogger.warn.bind(consoleLogger);


export default logger;

