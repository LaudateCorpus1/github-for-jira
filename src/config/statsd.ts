import { StatsCb, StatsD, Tags } from "hot-shots";
import { getLogger } from "./logger";
import { NextFunction, Request, Response } from "express";
import { EnvironmentEnum } from "./env";

const isTest = process.env.NODE_ENV === "test";

export const globalTags = {
	environment: isTest ? "test" : process.env.MICROS_ENV,
	environment_type: isTest ? "testenv" : process.env.MICROS_ENVTYPE,
	deployment_id: process.env.MICROS_DEPLOYMENT_ID || "1",
	region: process.env.MICROS_AWS_REGION || "localhost"
};

const logger = getLogger("config.statsd");

const statsd = new StatsD({
	prefix: "github-for-jira.",
	host: "platform-statsd",
	port: 8125,
	globalTags,
	errorHandler: (err) => {
		if (process.env.NODE_ENV !== EnvironmentEnum.development) {
			logger.warn(err, "Error writing metrics");
		}
	},

	mock: isTest
});

/**
 * High-resolution timer
 *
 * @returns {function(): number} A function to call to get the duration since this function was created
 */
function hrtimer() {
	const start = process.hrtime();

	return () => {
		const durationComponents = process.hrtime(start);
		const seconds = durationComponents[0];
		const nanoseconds = durationComponents[1];
		return seconds * 1000 + nanoseconds / 1e6;
	};
}

export const elapsedTimeMetrics = (
	req: Request,
	res: Response,
	next: NextFunction
) => {
	const elapsedTimeInMs = hrtimer();
	const { path, method } = req;
	const tags = { path, method };

	res.once("finish", () => {
		statsd.histogram("elapsedTimeInMs", elapsedTimeInMs(), tags);
	});

	next();
};

/**
 * Async Function Timer using Distributions
 */
export function asyncDistTimer(
	func: (...args: never[]) => Promise<unknown>,
	stat: string | string[],
	sampleRate?: number,
	tags?: Tags,
	callback?: StatsCb
) {
	return (...args: never[]): Promise<unknown> => {
		const end = hrtimer();
		const p = func(...args);
		const recordStat = () =>
			statsd.distribution(stat, end(), sampleRate, tags, callback);
		p.then(recordStat, recordStat);
		return p;
	};
}

export default statsd;
