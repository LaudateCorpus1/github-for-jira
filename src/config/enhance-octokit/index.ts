import statsd from "../statsd";
import { extractPath } from "../../backend/jira/client/axios";
import { GitHubAPI } from "probot";
import { metricHttpRequest } from "../metric-names";
import { getLogger } from "../logger";
import { Octokit } from "@octokit/rest";

const logger = getLogger("octokit");


const instrumentRequests = (octokit: GitHubAPI) => {
	octokit.hook.wrap("request", async (request, options) => {
		const requestStart = Date.now();
		let responseStatus = null;

		let response:Octokit.Response<any>;
		let error:any;
		try {
			response = await request(options);
			responseStatus = response.status;
			return response;
		} catch (err) {
			error = err;
			responseStatus = error?.responseCode;
			throw error;
		} finally {
			if(error || responseStatus < 200 || responseStatus >= 400) {
				logger.warn({request, response, error}, `Octokit error: failed request '${options.method} ${options.url}'`);
			}
			const elapsed = Date.now() - requestStart;
			const tags = {
				path: extractPath(options.url),
				method: options.method,
				status: responseStatus
			};

			statsd.histogram(metricHttpRequest().github, elapsed, tags);
		}
	});
};

/*
 * Customize an Octokit instance behavior.
 *
 * This acts like an Octokit plugin but works on Octokit instances.
 * (Because Probot instantiates the Octokit client for us, we can't use plugins.)
 */
export default (octokit: GitHubAPI): GitHubAPI => {
	instrumentRequests(octokit);
	return octokit;
};
