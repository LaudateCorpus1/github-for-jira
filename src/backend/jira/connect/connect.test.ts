import supertest from "supertest";
import express, { Express } from "express";
import setupFrontend from "../../app";
import { booleanFlag, BooleanFlags } from "../../../config/feature-flags";
import { when } from "jest-when";

jest.mock("../../../config/feature-flags");

describe("Connect", () => {
	let app: Express;

	const whenSignedInstallCallbacks = (value: boolean) =>
		when(booleanFlag).calledWith(
			BooleanFlags.USE_JWT_SIGNED_INSTALL_CALLBACKS,
			expect.anything()
		).mockResolvedValue(value);

	beforeEach(() => {
		// Defaults maintenance mode to true
		whenSignedInstallCallbacks(true);
		app = express();
	});

	describe("Frontend", () => {
		beforeEach(() => {
			app.use("/", setupFrontend({
				getSignedJsonWebToken: () => undefined,
				getInstallationAccessToken: () => undefined
			}));
		});

		describe("Atlassian Connect", () => {
			it("should return Atlassian Connect with signed-install:true when feature flag is off", () => {
				whenSignedInstallCallbacks(false);

				return supertest(app)
					.get("/jira/atlassian-connect.json")
					.expect(200)
					.then(response => {
						// removing keys that changes for every test run
						delete response.body.baseUrl;
						delete response.body.name;
						delete response.body.key;
						expect(response.body).toMatchSnapshot();
					})
			});

			it("should return Atlassian Connect with signed-install:true when feature flag is on", () => {

				whenSignedInstallCallbacks(true)

				return supertest(app)
					.get("/jira/atlassian-connect.json")
					.expect(200)
					.then(response => {
						// removing keys that changes for every test run
						delete response.body.baseUrl;
						delete response.body.name;
						delete response.body.key;
						expect(response.body).toMatchSnapshot();
					})
			});
		});

	});
});
