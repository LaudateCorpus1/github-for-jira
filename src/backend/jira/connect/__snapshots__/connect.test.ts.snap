// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connect Frontend Atlassian Connect should return Atlassian Connect with signed-install:true when feature flag is on 1`] = `
Object {
  "apiMigrations": Object {
    "gdpr": false,
    "signed-install": true,
  },
  "apiVersion": 1,
  "authentication": Object {
    "type": "jwt",
  },
  "description": "Connect your code and your project with ease.",
  "lifecycle": Object {
    "disabled": "/jira/events/disabled",
    "enabled": "/jira/events/enabled",
    "installed": "/jira/events/installed",
    "uninstalled": "/jira/events/uninstalled",
  },
  "modules": Object {
    "adminPages": Array [
      Object {
        "conditions": Array [
          Object {
            "condition": "user_is_admin",
          },
        ],
        "key": "gh-addon-admin",
        "location": "admin_plugins_menu/gh-addon-admin-section",
        "name": Object {
          "value": "Configure integration",
        },
        "url": "/jira/configuration",
      },
    ],
    "jiraBuildInfoProvider": Object {
      "homeUrl": "https://github.com/features/actions",
      "key": "github-actions",
      "logoUrl": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
      "name": Object {
        "value": "GitHub Actions",
      },
    },
    "jiraDeploymentInfoProvider": Object {
      "homeUrl": "https://github.com/features/actions",
      "key": "github-deployments",
      "logoUrl": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
      "name": Object {
        "value": "GitHub Actions",
      },
    },
    "jiraDevelopmentTool": Object {
      "application": Object {
        "value": "GitHub",
      },
      "capabilities": Array [
        "branch",
        "commit",
        "pull_request",
      ],
      "key": "github-development-tool",
      "logoUrl": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
      "name": Object {
        "value": "GitHub",
      },
      "url": "https://github.com",
    },
    "postInstallPage": Object {
      "conditions": Array [
        Object {
          "condition": "user_is_admin",
        },
      ],
      "key": "github-post-install-page",
      "name": Object {
        "value": "GitHub Configuration",
      },
      "url": "/jira/configuration",
    },
    "webSections": Array [
      Object {
        "key": "gh-addon-admin-section",
        "location": "admin_plugins_menu",
        "name": Object {
          "value": "GitHub",
        },
      },
    ],
  },
  "scopes": Array [
    "READ",
    "WRITE",
    "DELETE",
  ],
  "vendor": Object {
    "name": "Atlassian",
    "url": "https://atlassian.com",
  },
}
`;

exports[`Connect Frontend Atlassian Connect should return Atlassian Connect with signed-install:true when feature flag is off 1`] = `
Object {
  "apiMigrations": Object {
    "gdpr": false,
    "signed-install": true,
  },
  "apiVersion": 1,
  "authentication": Object {
    "type": "jwt",
  },
  "description": "Connect your code and your project with ease.",
  "lifecycle": Object {
    "disabled": "/jira/events/disabled",
    "enabled": "/jira/events/enabled",
    "installed": "/jira/events/installed",
    "uninstalled": "/jira/events/uninstalled",
  },
  "modules": Object {
    "adminPages": Array [
      Object {
        "conditions": Array [
          Object {
            "condition": "user_is_admin",
          },
        ],
        "key": "gh-addon-admin",
        "location": "admin_plugins_menu/gh-addon-admin-section",
        "name": Object {
          "value": "Configure integration",
        },
        "url": "/jira/configuration",
      },
    ],
    "jiraBuildInfoProvider": Object {
      "homeUrl": "https://github.com/features/actions",
      "key": "github-actions",
      "logoUrl": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
      "name": Object {
        "value": "GitHub Actions",
      },
    },
    "jiraDeploymentInfoProvider": Object {
      "homeUrl": "https://github.com/features/actions",
      "key": "github-deployments",
      "logoUrl": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
      "name": Object {
        "value": "GitHub Actions",
      },
    },
    "jiraDevelopmentTool": Object {
      "application": Object {
        "value": "GitHub",
      },
      "capabilities": Array [
        "branch",
        "commit",
        "pull_request",
      ],
      "key": "github-development-tool",
      "logoUrl": "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png",
      "name": Object {
        "value": "GitHub",
      },
      "url": "https://github.com",
    },
    "postInstallPage": Object {
      "conditions": Array [
        Object {
          "condition": "user_is_admin",
        },
      ],
      "key": "github-post-install-page",
      "name": Object {
        "value": "GitHub Configuration",
      },
      "url": "/jira/configuration",
    },
    "webSections": Array [
      Object {
        "key": "gh-addon-admin-section",
        "location": "admin_plugins_menu",
        "name": Object {
          "value": "GitHub",
        },
      },
    ],
  },
  "scopes": Array [
    "READ",
    "WRITE",
    "DELETE",
  ],
  "vendor": Object {
    "name": "Atlassian",
    "url": "https://atlassian.com",
  },
}
`;
